src/cascade/model/operations.py

```diff
@@ -21,68 +21,33 @@ def _var_grid_iterator(var_df):
-        meas_std = meas_std_full[["integrand_id", "covariate_id"]].drop_duplicates()
-        rate_value_full = sub_grid_df[(sub_grid_df.var_type == "mulcov_rate_value") & (sub_grid_df.covariate_id >= 0)]
-        rate_value = rate_value_full[["rate_id", "covariate_id"]].drop_duplicates()
-
-        categories = [rate_grids, meas_value, meas_std, rate_value]
-        category_cnt = sum([not test_kind.empty for test_kind in categories])
-        if category_cnt > 1:
-            raise RuntimeError(
-                f"The var table for smooth_id={smooth_id} should have only one kind "
-                f"but seems to have multiple kinds: rate? {not rate_grids.empty} meas value {not meas_value.empty} "
-                f"meas std? {not meas_std.empty} rate_value {not rate_value.empty}"
-            )
-        elif category_cnt == 0:
-            raise RuntimeError(
-                f"Cannot figure out why smooth_id={smooth_id} isn't one of "
-                f"a rate or a covariate multiplier")
-        # else: It's OK
-
-        if not rate_grids.empty:
-            node_id, rate_id = [int(rid) for rid in rate_grids.iloc[0]]
-            yield "rate", (smooth_id, rate_id, node_id), sub_grid_df
-
-        elif not meas_value.empty:
-            integrand_id, covariate_id = [int(meas_id) for meas_id in meas_value.iloc[0]]
-            yield "meas_value", (smooth_id, integrand_id, covariate_id), sub_grid_df
-
-        elif not meas_std.empty:
-            integrand_id, covariate_id = [int(std_id) for std_id in
-                                          meas_std.iloc[0]]
-            yield "meas_std", (smooth_id, integrand_id, covariate_id), sub_grid_df
-
-        elif not rate_value.empty:
-            rate_id, covariate_id = [int(rval_id) for rval_id in
-                                     rate_value.iloc[0]]
-            yield "rate_value", (smooth_id, rate_id, covariate_id), sub_grid_df
-
+    # groupby ignores every record where any of the keys are Null, so it
+    # only gets the rates.
+    for rate_index, sub_grid_df in var_df.groupby(["smooth_id", "rate_id", "node_id"]):
+        smooth_id, rate_id, node_id = rate_index
+        yield "rate", (smooth_id, rate_id, node_id), sub_grid_df
+
+    for integrand_index, sub_grid_df in var_df.groupby(["smooth_id", "integrand_id", "covariate_id"]):
+        smooth_id, integrand_id, covariate_id = integrand_index
+        var_type = sub_grid_df.iloc[0].var_type
+        if var_type == "mulcov_meas_value":
+            kind = "meas_value"
+        elif var_type == "mulcov_meas_std":
+            kind = "meas_std"
         else:
-            raise RuntimeError(f"Grid has unknown type for smooth_id={smooth_id}")
+            raise RuntimeError(f"Unknown var_type {var_type} in var table")
+        yield kind, (smooth_id, integrand_id, covariate_id), sub_grid_df
+
+    for alpha_index, sub_grid_df in var_df.groupby(["smooth_id", "rate_id", "covariate_id"]):
```

Comment:
Either I misunderstand what this is doing or I misunderstand how groupby works. Why do these groupbys not return overlapping results?

Reply:
Groupby ignores rows with Nulls in the given columns. That's why. I put a comment to that effect above.